File: lua/binary-search/binary-search.lua
return function(array, target)
	local l = 1
	local h = #array

	while l <= h do
		local m = l + math.floor((h - l) / 2)
		local item = array[m]

		if item == target then
			return m
		end

		if item < target then
			l = m + 1
		else
			h = m - 1
		end
	end

	return -1
end

File: zig/binary-search/binary_search.zig
pub fn binarySearch(comptime item: type, target: item, items: []const item) ?usize {
    if (items.len != 0) {
        var last = items.len - 1;
        var l: usize = 0;
        var h: usize = last;

        while (l <= h) {
            var m = (l + h) / 2;

            if (items[m] == target) {
                return m;
            }

            if (items[m] < target) {
                if (m == last) break;

                l = m + 1;
            } else {
                if (m == 0) break;

                h = m - 1;
            }
        }
    }

    return null;
}
